tests/server/server/test_callbacks.py

```diff
@@ -29,107 +31,112 @@ def test_json():
 
     new_response = callback_manager(request_set, response)
 
-    assert 10 == new_response['bodyAsJson']['count']
-    assert 'value' == new_response['bodyAsJson']['field']
-    assert 'count' in new_response['body']
+    assert 10 == new_response.bodyAsJson['count']
+    assert 'value' == new_response.bodyAsJson['field']
+    assert 'count' in new_response.body
 
     new_response = callback_manager(request, response)
-    assert 11 == new_response['bodyAsJson']['count']
-    assert 'value' == new_response['bodyAsJson']['field']
-    assert 'count' in new_response['body']
+    assert 11 == new_response.bodyAsJson['count']
+    assert 'value' == new_response.bodyAsJson['field']
+    assert 'count' in new_response.body
 
 
 def test_text():
     callback_manager.load('tests/server/server/callbacks')
     storage_manager.clear()
 
-    request = Request(method='get',
+    request = RequestBuilder.from_dict(dict(method='get',
                       host='api.com',
                       pathname='/text_counter',
                       query={'set': 10},
                       body='',
+                      protocol='http',
                       bodyAsJson={},
-                      headers={})
+                      headers={}))
 
-    response = Response(statusCode=200,
+    response = ResponseBuilder.from_dict(dict(statusCode=200,
                         body='Called',
                         bodyAsJson={},
-                        headers={})
+                        headers={}))
 
     new_response = callback_manager(request, response)
 
-    assert "Called 10 times" == new_response['body']
+    assert "Called 10 times" == new_response.body
 
 
 def test_json_full():
     callback_manager.load('tests/server/server/callbacks')
     storage_manager.clear()
 
-    request = Request(method='get',
+    request = RequestBuilder.from_dict(dict(method='get',
                       host='another.api.com',
                       pathname='/echo',
                       query={'message': 'Hello'},
                       body='',
+                      protocol='http',
                       bodyAsJson={},
-                      headers={})
+                      headers={}))
 
-    response = Response(statusCode=200,
+    response = ResponseBuilder.from_dict(dict(statusCode=200,
                         body='',
                         bodyAsJson={'field': 'value'},
-                        headers={})
+                        headers={}))
 
     new_response = callback_manager(request, response)
 
-    assert 'Hello' == new_response['bodyAsJson']['message']
-    assert 'value' == new_response['headers']['X-Echo-Header']
+    assert 'Hello' == new_response.bodyAsJson['message']
+    assert 'value' == new_response.headers['X-Echo-Header']
 
 def test_text_full():
     callback_manager.load('tests/server/server/callbacks')
     storage_manager.clear()
 
-    request = Request(method='post',
+    request = RequestBuilder.from_dict(dict(method='post',
```

Comment:
Now that `Request` etc. are proper classes, we shouldn't need "builder" and "writer" classes but could use instance or static methods in `Request` class directly I think.

Reply:
Definitely possible. The issue is that builder has some additional logic to transform fields, etc. But perhaps this should be offloaded to a constructor. Unless we want to keep the dataclass "pure" (no logic, just data).

Reply:
Yep, `Request.from_dict` just reads a lot better than `RequestBuilder.from_dict`. We could also make the dataclasses frozen to clearly indicate they're not supposed to be mutated.